IdentityCore/src/oauth2/aad_base/MSIDAADWebviewFactory.m

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDAADWebviewFactory.h"
+#import "MSIDWebviewConfiguration.h"
+#import "NSOrderedSet+MSIDExtensions.h"
+#import "MSIDWebWPJAuthResponse.h"
+#import "MSIDWebAADAuthResponse.h"
+#import "MSIDDeviceId.h"
+
+@implementation MSIDAADWebviewFactory
+
+- (NSMutableDictionary<NSString *,NSString *> *)authorizationParametersFromConfiguration:(MSIDWebviewConfiguration *)configuration requestState:(NSString *)state
+{
+    NSMutableDictionary<NSString *, NSString *> *parameters = [super authorizationParametersFromConfiguration:configuration
+                                                                                                 requestState:state];
+    
+    if (configuration.sliceParameters)
+    {
+        [parameters addEntriesFromDictionary:configuration.sliceParameters];
+    }
+
+    NSMutableOrderedSet<NSString *> *allScopes = parameters[MSID_OAUTH2_SCOPE].scopeSet.mutableCopy;
+    [allScopes addObject:MSID_OAUTH2_SCOPE_OPENID_VALUE];
+    
+    parameters[MSID_OAUTH2_SCOPE] = allScopes.msidToString;
+    parameters[MSID_OAUTH2_PROMPT] = configuration.promptBehavior;
+    
+    if (configuration.correlationId)
+    {
+        [parameters addEntriesFromDictionary:
+         @{
+           MSID_OAUTH2_CORRELATION_ID_REQUEST : @"true",
```

Comment:
I'm just curious why is this necessary?

Reply:
It was in x-plat libraries as well as our ADAL.

From MSDN, https://msdn.microsoft.com/en-us/library/mt739779.aspx

2.2.1.2 return-client-request-id
The return-client-request-id HTTP header is optional. This header is sent in the request and is used by the key provisioning server to determine whether to return the client-specified client-request-id in the server response. If present, the value of the return-client-request-id header MUST be "true".

The format of the return-client-request-id header, in ABNF, is as follows.

 return-client-request-id = "true"